### ISSUE TITLE
Big table OOMs on OPTIMIZE and DELETE

### ISSUE THREAD

**Issue by geronimogoemon** [ClickHouse/ClickHouse#42751]
Hello! I have a pretty big table partitioned by months and it seems one of the partitions has gotten so big that I can't do anything with it besides selecting data.

Running this query
```
SELECT partition_id, name,
    formatReadableSize(sum(bytes)) as size_human,
    sum(bytes) as size
    FROM system.parts
    where table = 'table_name' and partition_id = '202206'
GROUP BY table, partition_id, name
order by size desc
limit 10
```
I get 

|partition_id|name|size_human|size|
|------------|----|----------|----|
|202206|202206_1552078_2181294_49_2155140|126.09 GiB|135391399311|
|202206|202206_128269_1552076_85_2155140|119.82 GiB|128650542285|
|202206|202206_6_127740_45_2155140|107.48 GiB|115402133160|
|202206|202206_6_11_4_212|107.15 GiB|115050163853|
|202206|202206_2181300_2187592_4|11.04 GiB|11854580057|
|202206|202206_2187594_2193841_4|8.87 GiB|9528974398|
|202206|202206_2193865_2221791_4|7.15 GiB|7682434952|
|202206|202206_2223700_2225757_3|1.54 GiB|1652824250|
|202206|202206_2221810_2223688_3|1.49 GiB|1599094491|
|202206|202206_2225762_2226229_2|306.63 MiB|321521870|

When I try to run 
```
OPTIMIZE TABLE table_name PARTITION 202206;
```
I get an exception
```
DB::Exception: Memory limit (total) exceeded: would use 56.53 GiB (attempt to allocate chunk of 4624606 bytes), maximum: 56.53 GiB. OvercommitTracker decision: Memory overcommit isn't used. Waiting time or orvercommit denominator are set to zero..: (while reading column column_name): (while reading from part /home/clickhouse/data/store/c37/c3785c52-978a-4e7a-990f-c49c4ee4e917/202206_2225762_2226229_2/ from mark 0 with max_rows_to_read = 682): While executing MergeTreeSequentialSource. (MEMORY_LIMIT_EXCEEDED)
```
Moreover, when I tried deleting something from this partition the merge task kept failing due to OOM and restarting. I had to kill the mutation to stop it from running.

Unfortunately, I can't increase the memory limit in Clickhouse config since the box doesn't have more memory.

The table is an AggregatingMergeTree and I have all columns besides Primary Keys wrapped in SimpleAggregateFunction(anyLast) since the rows get updated with new values for different columns over time. There are 4 Primary Key columns and 516 non-PK columns and this particular partition has 1,7 billion rows. I am using Clickhouse version 22.6.2

Is there some special configuration or methodology that I am missing here which would enable performing normal maintenance on the table? Or is the table simply too big? 

Would having smaller partitions help? The data is stored over years and having weekly partitions would quickly result in over couple hundred of them which as far as I have seen in the recommendations is not advised.

**Comment by den-crane** [comment]
>The table is an AggregatingMergeTree and I have all columns besides Primary Keys wrapped in SimpleAggregateFunction(anyLast) since the rows get updated with new values for different columns over time. There are 4 Primary Key columns and 516 non-PK columns and this particular partition has 1,7 billion rows. I am using Clickhouse version 22.6.2

It sounds like a bug. Do you have VERYVERY long strings (String like 10 MB ).

22.6.2 is out of support. Try 22.8.8.3.
The size of the parts does not impact on memory used for merge.

Please share `select * from system.merges format Vertical` for this merge.

**Comment by Algunenano** [comment]
> The size of the parts does not impact on memory used for merge.

An exception is aggregation states, which seem to be the case here. A uniqueDistinctState() for example can, and will, run your DB if the source parts are big enough.

**Comment by geronimogoemon** [comment]
Apologies for a late reply. It took me quite some time to have time to update the Clickhouse server. I tested on
`Connected to ClickHouse server version 22.9.5 revision 54460.` and the same issues persists.

I wouldn't be surprised if having ~500 `SimpleAggregateFunction(anyLast, Nullable(Decimal(76, 18)))` columns was causing this problem since it seems unorthodox given my knowledge of Clickhouse. At the same time, the memory should only have to contain rows for a given PK which in this case is at most 5 so I am not sure what's the blocker here.

If the SimpleAggregateFunction is the perpetrator then having for example a ReplacingMergeTree in this case and the columns being just `Nullable(Decimal(76, 18)` would be a non-OOMing solution?

**Comment by llody55** [comment]
I also encountered the same problem on 23.3.2.37. I would like to know what your final solution was?
https://github.com/ClickHouse/ClickHouse/issues/57411

**Comment by den-crane** [comment]
@llody55 it's completely another problem.

The issue with memory consumption by merges is addressed in 23.5
